File: integrations/gateway/higress/continuation.go
package main

import (
	"strconv"
	"strings"

	"github.com/tidwall/gjson"
	"github.com/tidwall/sjson"
)

/*
 * THE RENDERING DIRECTIVE — how the runtime asks us to say no (OGR 1.3, 2026-08-28).
 *
 * ⚠️⚠️ **IT NEVER TURNS A REFUSAL INTO AN ALLOW.** `decision` is still `block` and
 * `Stops()` still answers true; this only selects WHICH refusal shape the caller gets.
 * The runtime derives the style from where the blocking findings sat — see the
 * platform's `policy-engine/continuation.ts` — because the position is what decides
 * whether a partial rendering is honest at all.
 *
 * ⚠️ **ABSENT IS THE NORMAL CASE AND MEANS "REFUSE AS YOU ALWAYS DID".** An older
 * runtime sends nothing, a runtime that could not name an honest shape sends nothing,
 * and both must land on the hard refusal. Every reader here therefore tests for a
 * style it KNOWS, never for "not empty" — an unrecognised style is a directive from a
 * newer runtime, and guessing at it is how a PEP forwards what it was told to remove.
 */

// continuationStyle values this build understands. Anything else falls back to a hard
// refusal, which is the strict side.
const (
	contDropCalls = "drop_calls"
	contWithhold  = "withhold"
	contAnswer    = "answer"
)

type continuation struct {
	Style  string
	Notice string
	Paths  []string
}

// Continuation returns the rendering directive, or nil when there is none or it names
// a style this build does not implement.
//
// ⚠️ A directive with a style we know but NO paths is rejected for the two styles that
// act on paths. A `drop_calls` naming nothing would append a notice saying an action
// was refused to a reply in which every action survived — the shape that reads as a
// working control and is not one.
func (v verdict) Continuation() *continuation {
	raw := v.root.Get("continuation")
	if !raw.Exists() {
		return nil
	}
	c := &continuation{
		Style:  raw.Get("style").String(),
		Notice: raw.Get("notice").String(),
	}
	for _, p := range raw.Get("paths").Array() {
		if s := p.String(); s != "" {
			c.Paths = append(c.Paths, s)
		}
	}
	if c.Notice == "" {
		return nil
	}
	switch c.Style {
	case contAnswer:
		return c
	case contDropCalls, contWithhold:
		if len(c.Paths) == 0 {
			return nil
		}
		return c
	default:
		return nil
	}
}

// StrippedPaths are the directive's paths in the form gjson/sjson address the body
// with — the same `payload.` strip the span applier does, for the same reason: the
// runtime speaks payload paths and this body has no `payload` wrapper.
func (c *continuation) StrippedPaths() []string {
	out := make([]string, 0, len(c.Paths))
	for _, p := range c.Paths {
		if s := stripPayloadPrefix(p); s != "" {
			out = append(out, s)
		}
	}
	return out
}

/*
 * DroppedOrdinals are the tool-call positions this directive refuses, read off the
 * trailing index of each path.
 *
 * ⚠️ The STREAM lane has no body to address, only the reassembled call list, so the
 * only thing a path can contribute there is WHICH call. A path whose tail is not an
 * integer contributes nothing rather than being guessed at — and the caller treats an
 * empty result as "cannot render survivors", which falls back to dropping the tail
 * whole. Guessing here would keep a refused call.
 */
func (c *continuation) DroppedOrdinals() []int {
	out := make([]int, 0, len(c.Paths))
	for _, p := range c.Paths {
		cut := strings.LastIndex(p, ".")
		if cut < 0 {
			return nil
		}
		n, err := strconv.Atoi(p[cut+1:])
		if err != nil || n < 0 {
			return nil
		}
		out = append(out, n)
	}
	return out
}

/*
 * withholdTexts replaces the text at each path with the notice and returns the new
 * body — the REQUEST-side continuation, and the only shape here that gives an agent
 * loop the same thing Claude Code's classifier gives it: a tool result that says why
 * it is not there, in a request that then proceeds normally.
 *
 * ⚠️⚠️ **NOTHING IS LEARNED INTO THE SESSION MAPPING, and that is the load-bearing
 * difference from `applySpans`.** A redaction's replacement is a placeholder the PEP
 * puts BACK on the way home; a withheld tool result must never come back. Feeding this
 * to the restore map would have the model's own reply rehydrate the exfiltrated page
 * we just removed — the control inverted, silently, on the return path.
 *
 * ⚠️ WHOLE-VALUE replacement, never a span. The runtime judged the entire tool result,
 * so offsets into it name nothing we want to keep; and a partial replacement would
 * leave the untrusted remainder in front of the model under a notice claiming it was
 * withheld.
 *
 * ⚠️ ALL OR NOTHING, like every other multi-path edit here. A path that does not
 * resolve means the body is not the one the runtime judged, and a partial withhold
 * forwards the content we refused.
 */
func withholdTexts(body string, paths []string, notice string) (string, bool) {
	if len(paths) == 0 {
		return body, false
	}
	for _, p := range paths {
		if gjson.Get(body, p).Type != gjson.String {
			return body, false
		}
	}
	out := body
	for _, p := range paths {
		next, err := sjson.Set(out, p, notice)
		if err != nil {
			return body, false
		}
		out = next
	}
	return out, true
}

File: integrations/gateway/java-proxy/core/src/main/java/com/openguardrails/ogr/Continuation.java
package com.openguardrails.ogr;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * HOW the runtime asks this proxy to say no (OGR 1.3).
 *
 * <h2>⚠️⚠️ It never turns a refusal into an allow</h2>
 *
 * {@code decision} is still {@code block} and the action still does not happen. This
 * selects WHICH SHAPE the refusal takes, because the shape the obvious implementation
 * uses — {@code finish_reason: "content_filter"} / {@code stop_reason: "refusal"} — is
 * the single token every agent harness treats as TERMINAL. One refused tool call in a
 * nine-step task ended the task. So the WHY moves off the finish reason and into two
 * places no loop branches on: an {@code x_ogr} body key and the notice the model reads.
 *
 * <h2>⚠️ Absent is the normal case and means "refuse as you always did"</h2>
 *
 * An older runtime sends nothing; a runtime that could not name an honest shape sends
 * nothing; both must land on the hard refusal. So every reader tests for a style it
 * KNOWS rather than for "not empty" — an unrecognised style is a directive from a newer
 * runtime, and guessing at it is how an enforcement point forwards what it was told to
 * remove.
 *
 * <p>The three styles, and which half of the step each acts on:
 *
 * <ul>
 *   <li>{@code drop_calls} — RESPONSE. Remove the named {@code payload.tool_calls.N}
 *       elements, keep the rest, append the notice. The calls the policy allowed still
 *       execute.
 *   <li>{@code withhold} — REQUEST. Replace the text at each path (a tool result) with
 *       the notice and FORWARD the request. The turn continues; the content does not
 *       reach the model.
 *   <li>{@code answer} — the notice is the whole reply, on a normal stop.
 * </ul>
 */
public final class Continuation {

    public static final String DROP_CALLS = "drop_calls";
    public static final String WITHHOLD = "withhold";
    public static final String ANSWER = "answer";

    public final String style;
    public final String notice;
    public final List<String> paths;

    Continuation(String style, String notice, List<String> paths) {
        this.style = style;
        this.notice = notice;
        this.paths = Collections.unmodifiableList(paths);
    }

    /**
     * The paths in the form a body resolver reads — the {@code payload.} prefix
     * stripped, because the runtime speaks payload paths and the body has no
     * {@code payload} wrapper.
     *
     * <p>A path that does not carry the prefix contributes nothing rather than being
     * guessed at.
     */
    public List<String> strippedPaths() {
        List<String> out = new ArrayList<String>(paths.size());
        for (String p : paths) {
            String s = Spans.stripPayloadPrefix(p);
            if (!s.isEmpty()) {
                out.add(s);
            }
        }
        return out;
    }

    /**
     * The tool-call ordinals this directive refuses, read off the trailing index of
     * each path.
     *
     * <p>⚠️ A STREAM has no body to address, only the reassembled call list, so the only
     * thing a path can contribute there is WHICH call. A path whose tail is not an
     * integer yields {@code null} for the whole set, and the caller then treats that as
     * "cannot render survivors" and drops the remainder whole. Guessing here would keep
     * a refused call.
     */
    public List<Integer> droppedOrdinals() {
        List<Integer> out = new ArrayList<Integer>(paths.size());
        for (String p : paths) {
            int cut = p.lastIndexOf('.');
            if (cut < 0) {
                return null;
            }
            try {
                int n = Integer.parseInt(p.substring(cut + 1));
                if (n < 0) {
                    return null;
                }
                out.add(Integer.valueOf(n));
            } catch (NumberFormatException e) {
                return null;
            }
        }
        return out;
    }

    @Override
    public String toString() {
        return "Continuation[" + style + " paths=" + paths + "]";
    }
}
